pkg/protocol/network.go

```diff
@@ -42,13 +37,6 @@ func newNetwork(protocol *Protocol, networkEndpoint network.Endpoint) *Network {
 // OnBlockReceived overwrites the OnBlockReceived method of the core protocol to filter out invalid blocks.
 func (n *Network) OnBlockReceived(callback func(block *model.Block, src peer.ID)) (unsubscribe func()) {
 	return n.Protocol.OnBlockReceived(func(block *model.Block, src peer.ID) {
-		// filter blocks from the future
```

Comment:
We need to have the wall clock check before a block reaches the `Protocol`. You moved the check now to the engine, but that is already to late, as the protocol also (or maybe only the protocol and not the engine) needs to have a perception of `current time`. 
If this is not the case, then a malicious actor could send a block with an arbitrary timestamp in the future and thus also commit to something far in the future and claim a massive weight (he just needs to be part of the committee once). Consequently, all nodes would see this chain as the heaviest and switch to it. It is therefore crucial to have a sense of `current time` also in the protocol when we sync chains.

Reply:
I moved it to the `ProcessResponse`, it's less hidden there and still should be early enough